### PR TITLE
Move JIT compilation backends to `wasmtime-jit`

### PR DESCRIPTION
**Pull request by alexcrichton**
This is an internal refactoring of the wasmtime crates to move actual
compilation of code from the `wasmtime-environ` crate to the
`wasmtime-jit` crate. Eventually it's hoped that all cranelift
dependencies can be removed from `wasmtime-environ`, but this is
intended to just be a first step for now.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -69,12 +69,7 @@ members = [
 
 [features]
 default = ["jitdump", "wasmtime/wat"]
-lightbeam = [
-    "wasmtime-environ/lightbeam",
-    "wasmtime-jit/lightbeam",
-    "wasmtime-wast/lightbeam",
-    "wasmtime/lightbeam",
-]
+lightbeam = ["wasmtime/lightbeam"]
```
Comment (bjorn3) [inline Cargo.toml:72]:
```suggestion
lightbeam = ["wasmtime/lightbeam", "wasmtime-jit/lightbeam"]
```

**Comment by sunfishcode** [comment]
wasmtime-environ is meant to be usable both for JIT and AOT use cases; can you discuss how AOT use cases fit within this refactoring?

**Comment by alexcrichton** [comment]
Oh sure yeah! So right now I imagine that if you're building an AOT backend then our current organization is probably not what you want. In theory you'd have to separate components, one for compilation and one for consumption of the compiled artifacts. The compiler would depend on cranelift at some point but the execution component would depend on `wasmtime-runtime` and doesn't want cranelift as a dependency. I'm hoping to refactor wasmtime internally to enable this use case.

The general idea is that you should be able to statically split the wasmtime world where the runtime half doesn't depend on the compiler half. Currently the "runtime half" is the `wasmtime-runtime` crate which depends on `wasmtime-environ` (for shared data structures), but `wasmtime-environ` unconditionally depends on codegen backends, making it unsuitable for this purpose.

I don't mean to conflate JIT/codegen/etc, and if you'd prefer I can extract all the codegen bits to something like a `wasmtime-codegen` crate or something like that.

**Comment by sunfishcode** [comment]
Yeah, splitting pieces of wasmtime-environ into separate crates sounds good to me. The "environ" in the name is meant to relate to the `FuncEnvironment` and `ModuleEnvironment` trait implementations that it contains, so if those move to a different crate, `wasmtime-environ` could benefit from a new name.

**Comment by alexcrichton** [comment]
I'm mostly just assuming the original intention for these crates, so I'm likely also misinterpreting things. Today though I think we still need to change something (`wasmtime-runtime` shouldn't transitively depend on cranelift/lightbeam). It sounds like you'd like to keep the code in `wasmtime-environ` separate from `wasmtime-jit`, so we'll need to split `wasmtime-environ`. The `wasmtime-runtime` crate would depend on the new crate that doesn't depend on cranelift. 

Does that sound right? Would you prefer that `wasmtime-environ` keep its name and a new crate is split out like `wasmtime-types` or `wasmtime-core` or something like that?

**Comment by alexcrichton** [comment]
I think I'll try to tackle this later.

**Pull request closed without merging**